app/main/views/agreements.py

```diff
@@ -237,3 +237,56 @@ def put_signed_framework_agreement_on_hold(agreement_id):
         return jsonify(message="Database Error: {0}".format(e)), 400
 
     return jsonify(agreement=framework_agreement.serialize())
+
+
+@main.route('/agreements/<int:agreement_id>/countersign', methods=['POST'])
+def countersign_agreement(agreement_id):
+    framework_agreement = FrameworkAgreement.query.filter(FrameworkAgreement.id == agreement_id).first_or_404()
+    framework_agreement_details = framework_agreement.supplier_framework.framework.framework_agreement_details
+
+    json_payload = get_json_from_request()
+    json_has_required_keys(json_payload, ['userId'])
+    updater_id = json_payload['userId']
+
+    updater_json = validate_and_return_updater_request()
+
+    if framework_agreement.status not in ['signed', 'on hold']:
+        abort(400, "Framework agreement must have status 'signed' or 'on hold' to be countersigned")
+
+    framework_agreement.signed_agreement_put_on_hold_at = None
+    framework_agreement.countersigned_agreement_returned_at = datetime.utcnow()
+
+    countersigner_details = {}
+    if framework_agreement_details:
+        if framework_agreement_details.get('countersignerName'):
+            countersigner_details.update({
+                'countersignerName': framework_agreement_details['countersignerName']
+            })
+        if framework_agreement_details.get('countersignerRole'):
+            countersigner_details.update({
+                'countersignerRole': framework_agreement_details['countersignerRole']
+            })
+
+    countersigner_details.update({'updatedById': updater_id})
```

Comment:
This is specifically the admin user who approved the agreement to be countersigned, and so `updatedById` seems too generic to me (this isn't just about updating a record, but specifically saying "this is good to go"). For uploaded agreements we have `agreedUserId` - this key might be better as `approvedByUserId` rather than `updatedById`.